### DISCUSSION TITLE
Add option to 'scramble / obfuscate / hash' ID's

### DISCUSSION THREAD

**Discussion by RobThree** [RobThree/HumanoID#5] [Ideas]
I'm thinking about a way to make the HumanoID's less predictable. Let me be very clear before we continue that I'm talking _less predictable_ here, not cryptographically secure or anything. Obfuscation, not encryption.

The way we'd achieve this is by providing an `HasherInterface` interface that only defines one method: `hash(int $id) : int`. This method should take an integer, shuffle/scramble or otherwise obfuscate the bits in such a way that when given a sequence of numbers each number isn't obviously predictable from the previous one. The function should be it's own inverse. Calling the method with it's own output should result in the original input.

We could accept and `HasherInterface` in the `HumanoID` constructor or default to a `NOPHasher` or `NOOPHasher` which does nothing (simply return the input). This way we won't break backwards compatibility (even though our downloads are still at zero as far as I can see 😂) but also provide the simplest and fastest default "hasher".

An _actual_ hasher could look like this (simple implementation for demonstration purposes):

```php
interface HasherInterface {
  public function hash(int $id) : int;
}

class DemoHasher implements HasherInterface{
  private int $seed;

  public function __construct(int $seed) {
    $this->seed = $seed;
  }

  public function hash(int $id) : int {
    return $this->seed ^ $id;
  }
}

// Create instance of DemoHasher 
$h = new DemoHasher (42069);

// Show hashes of values 1, 2 and 3
echo $h->hash(1) . PHP_EOL;
echo $h->hash(2) . PHP_EOL;
echo $h->hash(3) . PHP_EOL;

// And 'unhash' the values:
echo $h->hash(42068) . PHP_EOL;
echo $h->hash(42071) . PHP_EOL;
echo $h->hash(42070) . PHP_EOL;
```
Output:
```cmd
42068
42071
42070
1
2
3
```

This way we could provide one or two or a few 'hashers' (besides the default `NOPHasher`) that 'scamble' or 'shuffle' the bits of given id's in an deterministic, reversible, way that would help in 'obfuscating' sequential id's.

Our `HumanoID` constructor would get an extra argument:

```php
class HumanoID implements HumanoIDInterface
{
        public function __construct(
                array $wordSets,
                ?array $categories = null,
                ?string $separator = '-',
                ?WordFormatOption $format = null,
                ?HasherInterface $hasher = null
        ) {
                // ...
        }
// ...
}
```

And the `create()` and `parse()` methods would internally call the given (or default `NOPHasher` when `null`) hasher to '(de)obfuscate` given (humano)id's. Hashers we could provide out-of-the-box could be hashers based on [LFSR](https://en.wikipedia.org/wiki/Linear-feedback_shift_register)'s or what-have-you's with a known period or some other way of deterministically 'scrambling/shuffling' bits.

The default hasher implementation would look something like this:

```php
class NOPHasher implements HasherInterface {
  public function hash(int $id) : int {
    return $id;
  }
}
```

The `HumanoID`'s `create(...)` and `parse(...)` methods would then change as follows:

```php
// ...

    public function create(int $id): string
    {
        // Normalization / validation logic here...

        $id = $this->hasher->hash($id);  // Scramble bits
        // ...
    }

    public function parse(string $text): int
    {
        // Normalization / validation logic here...

        $id = $this->hasher->hash($id);  // Un-scramble bits
        // ...
    }

// ...
```

What are your thoughts on this @mallardduck and @MueR ?

**Comment by mallardduck** [comment]
I really dig this. Personally I know what my fav hasher will be is....

```php
function missy(int $id): int {
// PHP ints are 32-bits, so pad to at least that length
	$binaryInt = str_pad(decbin($id), 32, '0', STR_PAD_LEFT);
	$workIt = str_split($binaryInt, 1);
	$putMyThingDown = static function($thing) use (&$workIt) {
		array_unshift($workIt, $thing);
	};
	$putMyThingDown(array_pop($workIt));
	$flipIt = static function(&$it) {
		return array_map(static function($n) {
			$bool = filter_var($n, FILTER_VALIDATE_BOOLEAN);
			$invertBool = match ($bool) {
				true => false,
				false => true,
			};
			return (string) (int) $invertBool;
		}, $it);
	};
	$workIt = $flipIt($workIt);
	$workIt = array_reverse($workIt);
	return bindec(implode('', $workIt));
}
```

(just a rough 5 minute pass at the idea lol)

Can't ship this soon enough, will be very cool to have!

**Reply by RobThree** [reply]
I'm pretty sure it's a joke but in case it isn't: for the inputs `0..9` the output is:

```cmd
1
0
1
0
5
4
1
0
13
12
```

So I'm quite sure I see a little problem 🤓 😇

**Reply by mallardduck** [reply]
Yeah, was just a bit of a joke. 😆

**Reply by RobThree** [reply]
Thank God. I was about to send flowers and a card wishing you a speedy recovery 😉

**Reply by mallardduck** [reply]
Actually I think this joke has legs....I was just missing a step of padding the binary string to a minimum length. Padding the minimum `decbin` result to a length of 8 seems to fix all the lower value edge cases I found.

EDIT: I'm pretty sure given some actual testing behind it -rather than the meme hacking so far- that I could get this working correctly. The main "issue" seems to be ensuring PHP's `bindec` and `decbin` don't cause a footgun.

Primarily form integers that have different "binary string lengths", i.e. 5 = 101, becomes 110, then 001, then 100. However 2 = 10, becomes 01, then 10, then 01. But without the padding PHP will end up with `decbin(1) === 1` instead of `01`. So by padding with leading zeros to at least 8 it fixes all these edge cases.

**Reply by RobThree** [reply]
I think this is more up my alley 😆  I'll give it a try this weekend.

**Reply by mallardduck** [reply]
I tired to see what I can do to make it work with PHP's max int size, but it became quickly apparently that doing this would produce issues. I.e. padding to 64-bit (for most 64 bit systems) seems to allow for a creation of negative numbers. Playing with binary representation like this has been really neat at the very least even if it's not super viable.

**Reply by RobThree** [reply]
Since PHP has no support for unsigned ints the max. id we'll ever be able to support will be 2<sup>63</sup>-1 anyway. So I don't think we'll need to support anything larger than that (and it should suffice in most cases). All a hasher needs to do is do some bit-shuffling and avoid the MSB as far as I am concerned. Does that make sense?

**Comment by emmadesilva** [comment]
This doesn't really solve the problem in this thread, but wanted to write down an idea I had while playing around yesterday.

I originally used `create(1000 + $this->id)` to add some extra words. That could be used as a super-basic-not-at-all-cryptographically-secure way to "salt" values.

```php
// Psuedo code (untested)

// Setup
$superSecretSalt = 12345;
$userId = 42;

// Encode
$humanoID $generator->create($superSecretSalt + $userId)

// Decode
$userId = $generator->parse($humanoID) - $superSecretSalt;
```

**Reply by RobThree** [reply]
This would provide very basic 'obfuscation' but, although this particular example maybe being a little _too_ basic, I don't think we should aim for cryptographically secure anyway. It doesn't really make sense anyway. Actually, a HumanoID could be considered id obfuscation in itself. However, for known wordsets (like the two provided out-of-the-box) converting back to an integer would be trivial. I'd like to provide a way to add in a _teeny tiny little_ bit of obfuscation more and by providing an interface to do that we could easily provide one or more different ways to achieve this. I don't see an actual use case for 'cryptographically secure' HumanoID's but on the spectrum of no obfuscation at all to cryptographically secure there's a lot of in-between-obfuscations that could make sense. I guess some deterministic bit-shuffling would be somewhere on the middle of that line.

**Reply by emmadesilva** [reply]
Right! One thing I noticed and that bothered me a bit is that the HumanoIDs are sequential. I assume what the obfuscation you are trying here would solve that. Whereas my "salt" would just offset everything the same amount.

**Reply by RobThree** [reply]
> HumanoIDs are sequential

Not, they're not. They're not sequential nor random. They are whatever your input is. If you feed it sequential ID's then the HumanoID's will be "sequential" too. If you feed it random ID's the output will be random as well. But HumanoID's aren't inherently ordered (you _could_ argue they're wordlist-ordered if you feed sequential integers).

But the most common use-case will be to replace (usually sequential) autoincremented database keys with HumanoID's in urls for example. And that is where HumanoID's could aid in 'obfuscating' the ID's and where an "Hasher" could help obfuscate things a bit more by shuffling some bits to obfuscate the order a little when the wordlist is known.

**Reply by emmadesilva** [reply]
Sorry, you are exactly right.

My use case is exactly as you explained, replacing sequential autoincremented database keys with HumanoIDs.

**Comment by RobThree** [comment]
After giving it some more thought, I think the proposed interface:

```php
interface HasherInterface {
  public function hash(int $id) : int;
}
```

Is a little too limiting; this only allows functions that are reversible by calling the same method again. But there are functions that are reversible by doing operations A->B->C initially and then C->B->A to revert. I think a better approach would be:

```php
interface ShufflerInterface {
  public function shuffle(int $id) : int;
  public function unshuffle(int $id) : int;
}
```

I don't think hashing nor shuffling is the correct term here, I'll sit on that one too for a bit.

**Reply by emmadesilva** [reply]
> I don't think hashing nor shuffling is the correct term here, I'll sit on that one too for a bit.

Why not scramble/obfuscate, like in the discussion title?

Good call on extending the interface, I think that makes it easier to use.

**Reply by RobThree** [reply]
Agree, it will probably end up as (a variation of) scramble/obsfuscate.

**Comment by RobThree** [comment]
Although I have some ideas (and even implemented some a long time ago) myself, I found some 'bit shuffling methods' [here](https://blog.jiayu.co/2018/09/methods-for-obfuscating-sequential-ids/) and [here](https://ericlippert.com/2013/11/14/a-practical-use-of-multiplicative-inverses/). We have some 'unusual' requirements though; we want to obfuscate a sequence, but we don't want simple ID's like `17` ending up as `2398724398` before converting to HumanoID because that will result in an unnecessarily long string. On the other hand: some people may not care about long(er) strings.

I think I found at least one candidate which we could use as a '3rd party dependency': [Optimus](https://github.com/jenssegers/optimus). I guess the way to go would be to create an `HumanoID.Optimus` package or something which would just reference the `Optimus` package (and HumanoID ofcourse) and implement the `HumanoID` 'obfuscator interface` using Optimus' encode/decode methods.

**Reply by mallardduck** [reply]
Agreed. Once the interface is published in this package, then any obfuscator that uses 3rd party libraries could be built as a "plugin library". Seems like a solid path.

**Comment by mallardduck** [comment]
hiya @RobThree and @caendesilva - I decided to hop on this and take a shot at an initial implementation...please check it out when you can. Very early stages, but mostly just taking a lot of what we discussed here and seeing what sticks.

https://github.com/RobThree/HumanoID/pull/13

**Reply by emmadesilva** [reply]
Taking a look! Glad to hear from you, doing code reviews with you guys is really fun!

**Reply by RobThree** [reply]
> Taking a look! Glad to hear from you, doing code reviews with you guys is really fun!

Are you kidding me? As you posted this I was writing [the exact same thing](https://github.com/RobThree/HumanoID/pull/13#issuecomment-1159107135) 🤣

**Comment by RobThree** [comment]
Just merged #13
